### PR TITLE
Bug 1920248: fix pipeline rerun when pipelineSpec is embedded

### PR DESCRIPTION
**Pull request by nemesis09**
Fixes:
https://issues.redhat.com/browse/ODC-5395

Analysis / Root cause:
While re-running the pipelineRun with embedded pipelinespec causes the UI to crash.

Solution Description:
update util getPipelineRunData and kebab actions to handle the scenario when pipeline is passed in pipelineSpec instead of a reference in pipelineRef

Screen shots / Gifs for design review:
pipelineSpec

Test Coverage:
Screenshot from 2021-01-30 20-55-07

Browser Conformance:
- [x] Firefox
- [ ] Chrome
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/pipelines-plugin/src/utils/pipeline-actions.tsx [hunk 1]
```diff
@@ -43,8 +43,9 @@ export const reRunPipelineRun: KebabAction = (kind: K8sKind, pipelineRun: Pipeli
   labelKey: 'pipelines-plugin~Rerun',
   callback: () => {
     const namespace = _.get(pipelineRun, 'metadata.namespace');
-    const pipelineRef = _.get(pipelineRun, 'spec.pipelineRef.name');
-    if (namespace && pipelineRef) {
+    const pipelineRef = pipelineRun.spec.pipelineRef?.name;
+    const { pipelineSpec } = pipelineRun.spec;
+    if (namespace && (pipelineRef || pipelineSpec)) {
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/utils/pipeline-actions.tsx:48]:
```suggestion
    const { pipelineSpec, pipelineRef } = pipelineRun.spec;
    if (namespace && (pipelineRef?.name || pipelineSpec)) {
```

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts [hunk 2]
```diff
@@ -84,13 +91,18 @@ export const getPipelineRunData = (
...
+          'tekton.dev/pipeline': pipelineName,
+        },
+      ),
     },
     spec: {
       ...(latestRun?.spec || {}),
-      pipelineRef: {
+      pipelineRef: latestRun.spec.pipelineRef && {
         name: pipelineName,
       },
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts:107]: Perhaps avoid adding it if it's not there.
```suggestion
      ...(latestRun.spec.pipelineRef && { pipelineRef: {
        name: pipelineName,
      } }),
```

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts [hunk 3]
```diff
@@ -43,6 +43,12 @@ export const migratePipelineRun = (pipelineRun: PipelineRun): PipelineRun => {
   return newPipelineRun;
 };
 
+export const getPipelineName = (pipeline?: Pipeline, latestRun?: PipelineRun) =>
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts:46]: Needs tests! 😜

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts [hunk 4]
```diff
@@ -43,6 +43,12 @@ export const migratePipelineRun = (pipelineRun: PipelineRun): PipelineRun => {
...
+  pipeline
+    ? pipeline.metadata.name
+    : latestRun.spec.pipelineRef?.name ??
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts:49]: if you don't have a pipeline or a latestRun, you'll get an NPE 🙂  -- when you write tests you will see the logic shift.

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts [hunk 5]
```diff
@@ -68,6 +74,7 @@ export const getPipelineRunData = (
     {},
     pipeline?.metadata?.annotations,
     latestRun?.metadata?.annotations,
+    !latestRun?.metadata?.annotations?.preferredName && { preferredName: pipelineName },
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts:77]: Only real question here is do we want to call this something else.

cc @vdemeester 

Essentially what's happening here is we have a PipelineRun without a reference to a Pipeline. However, we build our PipelineRun names off of the Pipeline name + 6 random characters. Without that, we sorta need some clean reference otherwise we'll end up with PipelineRun + 6 random characters + 6 random... etc. So we are going with the first one we set in the annotations, "preferredName" was off the top of my head. Thoughts?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/packages/pipelines-plugin/src/utils/pipeline-actions.tsx (hunk 1 above)
Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/utils/pipeline-actions.tsx]: done

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts (hunk 2 above)
Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts]: done

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts (hunk 3 above)
Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts]: added tests

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts (hunk 4 above)
Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts:49]: We wouldn't reach the call to `getPipelineName` if pipeline and `latestRun` both are absent, since it is handled at the very beginning of the definition of `getPipelineRunData`

Comment (andrewballantyne) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts:49]: This function knows nothing of where it was called. If you write fully comprehensive tests for this function, you'd pass null for both values and see it crashes. Please add more tests.

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/__tests__/utils.spec.ts [hunk 6]
```diff
@@ -239,3 +240,36 @@ describe('convertPipelineToModalData', () => {
     expect(workspaces.filter((workspace) => workspace.type === 'PVC')).toHaveLength(3);
   });
 });
+
+describe('getPipelineName', () => {
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/__tests__/utils.spec.ts:244]: Initial look, you're missing a couple tests here:

- what if nothing is passed?
- what if nulls are passed for one / both?
- what if both pipeline AND pipeline run are passed?

Remember, Unit Tests are to catch all uses of a function, not just the ones from the expected path.

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts (hunk 5 above)
Comment (vdemeester) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts]: Do we want to *namespace* the annotation name, like `io.openshift.pipeline/preferredName` ?
Otherwise, seems good to me.

What happens if the pipelinerun doesn't have this annotation ? Will it then add the annotation for the next one ?

Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts]: Thanks for the input @vdemeester 
yes, if the pipeline run doesn't have this annotation, it will be added to the next one.
I think we could have the annotation namespaced.  cc @andrewballantyne

Comment (andrewballantyne) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts]: Yup. Namespacing it sounds like a good idea

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/__tests__/utils.spec.ts (hunk 6 above)
Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/__tests__/utils.spec.ts:244]: added tests for following scenarios
- nothing being passed to the function, i.e. both values falsy
- both pipeline and pipeline run being passed to the function

added a check in the function to return null if both arguments are falsy. So do we need another test for null values, since it will treat null as falsy value?

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts (hunk 4 above)
Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts:52]: added a check in the function to handle the scenario when both arguments are falsy

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts [hunk 7]
```diff
@@ -43,6 +43,16 @@ export const migratePipelineRun = (pipelineRun: PipelineRun): PipelineRun => {
...
 
+export const getPipelineName = (pipeline?: Pipeline, latestRun?: PipelineRun): string => {
+  if (!pipeline && !latestRun) {
+    return null;
+  }
+  return pipeline
+    ? pipeline.metadata.name
+    : latestRun.spec.pipelineRef?.name ??
+        (latestRun.metadata.annotations?.preferredName || latestRun.metadata.name);
+};
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts:54]:
```suggestion
export const getPipelineName = (pipeline?: Pipeline, latestRun?: PipelineRun): string => {
  if (pipeline) {
    return pipeline.metadata.name
  }
  if (pipelineRun) {
    return latestRun.spec.pipelineRef?.name ??
        (latestRun.metadata.annotations?.preferredName || latestRun.metadata.name)
  }
  return null;
};
```
Nit: You could write this in a fashion where it's flatter and easier to compartmentalize the logic.

Not required, just a thought.

**Review by andrewballantyne — COMMENTED** [review]
Still need the annotation change, and then I can run it through the paces.

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts (hunk 7 above)
Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts]: this does seem to be easier to read and compartmentalize the logic.
updated.

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts [hunk 8]
```diff
@@ -18,6 +18,9 @@ import { CREATE_PIPELINE_RESOURCE, initialResourceFormValues } from './const';
 import { CommonPipelineModalFormikValues, PipelineModalFormResource } from './types';
 import { TektonResourceLabel, VolumeTypes } from '../../const';
 
+// Annotation for referencing pipeline name in case of PipelineRun with no reference to a Pipeline (embedded pipeline)
+export const preferredNameAnnotation = 'io.openshift.pipeline/preferredName';
```
Comment (karthikjeeyar) [inline frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts:22]: We have an  **started-by** annotations already, it is like this  `pipeline.openshift.io/started-by`. We should be consistent here
```suggestion
export const preferredNameAnnotation = 'pipeline.openshift.io/preferredName';
```
and also move this to `src/components/pipelines/const.ts`

File: frontend/packages/pipelines-plugin/src/test-data/pipeline-data.ts [hunk 9]
```diff
@@ -25,6 +25,7 @@ export enum PipelineExampleNames {
   CONDITIONAL_PIPELINE = 'conditional-pipeline',
   INVALID_PIPELINE_MISSING_TASK = 'missing-task-pipeline',
   INVALID_PIPELINE_INVALID_TASK = 'invalid-task-pipeline',
+  EMBEDDED_PIPELINESPEC = 'embedded-spec-pipeline-run',
```
Comment (karthikjeeyar) [inline frontend/packages/pipelines-plugin/src/test-data/pipeline-data.ts:28]: nit:
```suggestion
  EMBEDDED_PIPELINE_SPEC = 'embedded-pipeline-spec',
```

File: frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts (hunk 8 above)
Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts]: done

File: frontend/packages/pipelines-plugin/src/test-data/pipeline-data.ts (hunk 9 above)
Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/test-data/pipeline-data.ts]: done

**Comment by spadgett** [comment]
/override ci/prow/ceph-storage-plugin

The ceph-storage-plugin job should only have run on PRs with changes to `frontend/packages/ceph-storage-plugin`. It failed due to an unrelated CI config change. Overriding status on this PR.

File: frontend/packages/pipelines-plugin/src/test-data/pipeline-data.ts [hunk 10]
```diff
@@ -1885,6 +1886,115 @@ export const pipelineTestData: PipelineTestData = {
       },
     },
   },
+  [PipelineExampleNames.EMBEDDED_PIPELINE_SPEC]: {
+    dataSource: 'embedded-pipelineSpec',
+    pipeline: {
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/test-data/pipeline-data.ts:1891]:
```suggestion
    pipeline: null,
```
You're in conflict here. You crafted a Pipeline for a test chunk that says it has an embedded pipeline 🙂

**Review by andrewballantyne — COMMENTED** [review]
Code looks pretty good. Going to test it on a cluster now before approving.

File: frontend/packages/pipelines-plugin/src/test-data/pipeline-data.ts (hunk 10 above)
Comment (nemesis09) [reply frontend/packages/pipelines-plugin/src/test-data/pipeline-data.ts]: it indeed is conflicting. Thanks for the catch.
updated the test data

**Review by andrewballantyne — APPROVED** [review]
Nice work with the tests Aritra!

**Comment by nemesis09** [comment]
> Nice work with the tests Aritra!

thank you :)

**Comment by karthikjeeyar** [comment]
Thanks @nemesis09  Verified locally, works fine. 
/lgtm

**Comment by karthikjeeyar** [comment]
@nemesis09 Since pipeline has its own package in 4.7, we need to manually cherrypick this code to 4.6
